testrepository/tests/test_testcommand.py

```diff
@@ -435,15 +604,90 @@ def test_run_tests_with_instances(self):
+            ('values', [('running', _u('foo 1 P1'))]),
+            ('popen', (_u('foo 1 P1'),), {'shell': True, 'stdin': -1, 'stdout': -1})
+            ]), Equals([
+            ('popen', ('probe',), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', _u('foo 1 P1'))]),
+            ('popen', (_u('foo 1 P1'),), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('values', [('running', _u('foo 2 P2'))]),
+            ('popen', (_u('foo 2 P2'),), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ])))
+
+    def test_run_tests_with_profiles_and_instances(self):
+        # when there are profiles and instance_execute, run_tests runs
+        # profile cmds in instances.
+        self.set_config(dedent("""\
+            [DEFAULT]
+            test_command=foo $IDLIST $PROFILE
+            list_profiles=probe
+            instance_provision=p -c $INSTANCE_COUNT $PROFILE
+            instance_execute=q $INSTANCE_ID -- $COMMAND
+            """))
+        ui, command = self.get_test_ui_and_cmd2()
+        ui.proc_outputs = [b'P1 P2', b'I1', b'', b'I2', b'']
+        self.useFixture(command)
+        fixture = self.useFixture(command.get_run_command(
+            test_ids={'1': {'profiles': ['P1']}, '2': {'profiles': ['P2']}}))
+        procs = fixture.run_tests()
+        self.expectThat(ui.outputs, MatchesAny(Equals([
+            ('popen', ('probe',), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', 'p -c 1 P1')]),
+            ('popen', ('p -c 1 P1',), {'shell': True, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', _u('q I1 -- foo 1 P1'))]),
+            ('popen', (_u('q I1 -- foo 1 P1'),), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('values', [('running', 'p -c 1 P2')]),
+            ('popen', ('p -c 1 P2',), {'shell': True, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', _u('q I2 -- foo 2 P2'))]),
+            ('popen', (_u('q I2 -- foo 2 P2'),), {'shell': True, 'stdin': -1, 'stdout': -1})
+            ]), Equals([
+            ('popen', ('probe',), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', 'p -c 1 P2')]),
+            ('popen', ('p -c 1 P2',), {'shell': True, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', _u('q I1 -- foo 2 P2'))]),
+            ('popen', (_u('q I1 -- foo 2 P2'),), {'shell': True, 'stdin': -1, 'stdout': -1}),
+            ('values', [('running', 'p -c 1 P1')]),
+            ('popen', ('p -c 1 P1',), {'shell': True, 'stdout': -1}),
+            ('communicate',),
+            ('values', [('running', _u('q I2 -- foo 1 P1'))]),
+            ('popen', (_u('q I2 -- foo 1 P1'),), {'shell': True, 'stdin': -1, 'stdout': -1})
+            ])))
```

Comment:
IMO this is getting a little out of hand :wink: 

I think the intent of the test is obscured by the matcher-of-matcher-of-list-of-tuple-of-dict data structure. Do you think it's worth abstracting these away, so the test could use more expressive names?